src/twisted/python/compat.py

```diff
@@ -852,6 +852,37 @@ def _get_async_param(isAsync=None, **kwargs):
 
 
 
+def _pypy3BlockingHack():
+    """
+    Work around U{this pypy bug
+    <https://bitbucket.org/pypy/pypy/issues/3051/socketfromfd-sets-sockets-to-blocking-on>}
+    by replacing C{socket.fromfd} with a more conservative version.
+    """
+    try:
+        from fcntl import fcntl, F_GETFL, F_SETFL
+    except ImportError:
+        return
+    if not (_PY3 and _PYPY):
+        return
+
+    def fromFDWithoutModifyingFlags(fd, family, type, proto=None):
+        passproto = [proto] * (proto is not None)
+        flags = fcntl(fd, F_GETFL)
+        try:
+            return realFromFD(fd, family, type, *passproto)
+        finally:
+            fcntl(fd, F_SETFL, flags)
+    realFromFD = socket.fromfd
+    if realFromFD.__name__ == fromFDWithoutModifyingFlags.__name__:
+        return
+    socket.fromfd = fromFDWithoutModifyingFlags
```

Comment:
So... personally I have experienced cases where twisted has monkey patched globally a thing and that monkey patch *introduced* bugs in non-twisted code...

I realize how tempting this is over instead just changing code to use a level of indirection, since I presume fromfd is reasonably widely used but... severe :(

Tell me to go away though I guess, and that this is the best way.

Reply:
This is the best way, at least in this specific case.

- It's only monkeypatched on pypy3
- The behavior on pypy3 is buggy as heck and this will fix it for other programs as well.
- The monkeypatch is harmless (even if the monkeypatch were to get invoked on a different cpython, it would at worst be slightly inefficient unless someone is doing something *truly* batty with threads).
- I expect this to be relatively short-lived; it links to the bug, the bug is super straightforward and has a reproducer that's only a half a dozen lines.  The other bug I filed today, Armin has already fixed.